### ISSUE TITLE
Idea: pretrain a ModernBERT-style factory encoder via masked-tile modelling on scraped human blueprints

### ISSUE THREAD

**Issue by beyarkay** [beyarkay/factorion#160]
## Idea (speculative): pretrain a factory *foundation* encoder with masked-tile modelling on scraped human blueprints

A research bet, not a committed task. Filing it so we don't lose it.

### TL;DR

Treat a factory grid like "text". Bulk-download a large corpus of **human-made
Factorio blueprints**, mask out N% of tiles, and train an encoder to predict the
masked tiles (BERT/ModernBERT-style masked "language" modelling). That gives us a
strong, self-supervised **prior over what real factories look like**. *Then* bolt a
task head on top of the pretrained encoder and SFT/RL it to actually solve factories.

This is the bitter-lesson move: instead of hand-authoring synthetic lessons
(`build_factory` + `blank_entities`), let scale do the work — pretrain on ~all the
human factory layouts that exist, which is a much bigger and more diverse signal than
our generators produce.

### How it slots into the current pipeline

Today: **synthetic SFT** (`build_factory`/`blank_entities` pairs) → **PPO finetune**.

Proposed: add a **stage 0** in front:

```
stage 0  MLM pretrain encoder on scraped human blueprints   (self-supervised, no labels)
stage 1  SFT task head on synthetic pairs                    (our existing sft.py)
stage 2  PPO finetune                                        (our existing ppo.py)
```

The nice part: stages 1–2 already exist, and the encoder is already a shared,
swappable component (`AgentCNN`'s dynamic conv encoder). MLM pretraining just produces
better *initial encoder weights* than random init. The masked-tile objective is a
near-perfect fit for our representation — factory state is already a 3D
channels×W×H tensor with semantic channels, so "predict the masked tile" = predict the
per-cell entity/direction/item over masked grid positions. This is essentially the
`blank_entities` objective but (a) self-supervised on real data and (b) without needing
a known-correct completion, since the blueprint *is* the ground truth.

It may also directly attack the things we've been fighting:
- **#106** (dir_acc plateaus ~91%, "irreducible label noise") and the multi-path
  ambiguity we keep hitting — a model with a real prior over human layouts should
  resolve more of that ambiguity the way a human would.
- **#113 / #130** (lesson quality, best-of-N data filtering) — real blueprints sidestep
  "are our synthetic lessons good examples?" entirely.

### Data feasibility — verified, blueprints are trivially easy to bulk-download

A blueprint is just a base64-encoded, zlib-compressed JSON string — kilobytes each.
The dominant repo, **Factorio Prints**, runs on a *public* Firebase Realtime Database
(project name is the misspelled `facorio-blueprints`). REST: append `.json` to any
path, no API key, public read.

**I confirmed this is live right now** (June 2026): a shallow GET against
`https://facorio-blueprints.firebaseio.com/blueprintSummaries.json?shallow=true`
returns real keys, e.g. `{"-KYbIg8cULtwPnFfpb7b":true, "-KYcDgD30qUnuPBXl6kW":true, …}`
— not a permission-denied error. The historical access pattern (their own issue #27)
was: one GET of the whole `blueprintSummaries` index, filter client-side, then N calls
to `/blueprints/{id}.json` for payloads. **factorio.school** is the same codebase
pointing at the same backend, so scraping one gets both.

- Don't even need to write the scraper: the `factorio-prints-browser` mod ships a Node
  script (`fetch-blueprints.mjs`, `--limit=N --sort=date`) that walks the live Firebase.
- Other sources: **factorioblueprints.tech** (open source, `barthuijgen/factorio-sites`,
  has its own pullable DB), factoriocodex.com, plus the pastebin/gist/Reddit long tail.

**Scale:** order tens of thousands of blueprints; at a few KB of string each the entire
corpus is low hundreds of MB even with metadata + preview images. No technical ceiling —
grabbable in an afternoon. **Only real constraint is courtesy**: it's a hobbyist Firebase
that's already had bandwidth/cost problems (the reason they moved to pagination), so
rate-limit ourselves rather than hammer it.

**Decode pipeline:** base64-decode → strip the leading version byte → `zlib.inflate` →
JSON of entities. We then rasterise that entity JSON onto our existing grid-tensor
representation (same encoding the env already uses).

### Saves/worlds: the opposite situation (not worth chasing now)

No central DB of full save games. Saves are binary `.zip`, 5–50+ MB, version-locked,
shared ad hoc (Dropbox/Mega/Drive/forum). Bulk-collecting played worlds is a real,
messy crawling project with low yield. ("Map exchange strings" are just gen seeds, not
played factories.) **But** the user's note stands: we can cheaply *manufacture*
blueprints from our own/borrowed world saves (select-area → blueprint), so saves can be
a manual top-up source, not a scraping target.

### Open questions / risks

- **License.** Blueprints are user-submitted with no clear license / no blanket grant
  for redistribution or training use. This is unsettled and ours to sort out before we
  publish anything trained on it. (Pretraining for internal research is lower-risk than
  redistributing the corpus or a model derived from it.)
- **Layout, not behaviour.** Blueprint strings encode entity placement, *not*
  throughput. The MLM objective only needs structure, so this is fine for stage 0 — but
  if we want functional labels we compute them ourselves (we already have
  `simulate_throughput`). The encoder learns "what plausible factories look like", not
  "what works"; stages 1–2 supply the works-or-not signal.
- **Masking design for a grid.** Open questions: mask whole entities vs individual
  channels of a cell? Span/region masking (mask a contiguous patch) to force long-range
  structure, à la ModernBERT? What mask ratio? How to handle the huge empty-tile class
  imbalance (most cells are empty) so the model doesn't just learn "predict empty"?
- **Vocabulary / real-game entities.** Real blueprints contain far more entity types,
  recipes, and modules than our current env models. We either (a) restrict the corpus to
  the entity subset we support, or (b) expand our channel vocabulary to cover them
  (bigger, but more transfer). Filtering by entity set is the easy first cut.
- **Distribution gap.** Human blueprints are large, full-game designs; our tasks are
  tiny grids. May need to crop/tile big blueprints into patches that match our grid size.
- **Does it actually beat synthetic-only SFT?** The whole bet. Needs an ablation:
  MLM-pretrained encoder → SFT vs random-init → SFT, on our existing eval.

### Prior art

- **Factorio Learning Environment** (arXiv 2503.09617) deliberately does *not* use
  scraped human designs — it's a programmatic API + procedural eval. So scraped
  blueprints as a *training* corpus is a relatively unexplored angle.
- "Towards Automatic Design of Factorio Blueprints" (arXiv 2310.01505, ModRef 2023) —
  treats it as combinatorial optimisation (bin-packing + routing + network design), not
  learned-prior. Useful framing for what "solve" means.
- `piebro/factorio-blueprint-visualizer` and an existing SDXL finetune on
  blueprint-rendered images show the corpus is already being scraped/used elsewhere —
  reusable tooling for the decode/render side.

### Concrete first spikes (cheap, de-risk before committing)

1. **Scraper spike** (~half a day): pull `blueprintSummaries` index, fetch a few hundred
   `/blueprints/{id}.json`, decode (base64 → strip version → inflate), confirm we can
   rasterise onto our grid tensor. Rate-limited, polite.
2. **Corpus stats:** how many decode cleanly? entity-type histogram → what fraction fall
   inside our current vocabulary? typical sizes → tiling strategy.
3. **MLM head spike:** add a masked-tile head on `AgentCNN`'s encoder, train on the
   scraped corpus, watch reconstruction accuracy. No task head yet.
4. **Ablation:** MLM-pretrained encoder → existing SFT vs random-init → existing SFT, on
   the current eval. This is the go/no-go.

### Caveats to re-verify before building

- Firebase read rules can be tightened unilaterally by the owner — re-check the shallow
  GET before each big pull.
- Get an actual live count + entity histogram before assuming corpus size/coverage.
- Settle the licensing question before any redistribution.
